src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java

```diff
@@ -57,29 +58,294 @@
 /** Syntax node for a function call expression. */
 public final class FuncallExpression extends Expression {
 
+  /** A value class to store {@link ParamType} metadata to avoid using Java proxies. */
+  public static final class ParamTypeDescriptor {
+
+    private final Class<?> type;
+    private final Class<?> generic1;
+
+    private ParamTypeDescriptor(Class<?> type, Class<?> generic1) {
+      this.type = type;
+      this.generic1 = generic1;
+    }
+
+    /** @see ParamType#type() */
+    public Class<?> getType() {
+      return type;
+    }
+
+    /** @see ParamType#generic1() */
+    public Class<?> getGeneric1() {
+      return generic1;
+    }
+
+    static ParamTypeDescriptor of(ParamType paramType) {
+      return new ParamTypeDescriptor(paramType.type(), paramType.generic1());
+    }
+  }
+
+  /** A value class for storing {@link Param} metadata to avoid using Java proxies. */
+  public static final class ParamDescriptor {
```

Comment:
Could you please move this class to the top level? Long class definitions, even if they're trivial, make reading the containing class a bit more daunting.

Reply:
absolutely, I've used a nested class because MethodDescriptor is nested, but given how big all of these classes are, I agree that separating them makes code more readable :)